PythonTool/TagExpansion.py

```diff
@@ -25,104 +24,171 @@ def parse_arguments():
+           The type of MOAB element from which to extract the tag list,
+           represented by an integer.
 
     Returns:
     ________
-       element_list: List
-           The list of all specific elements in the mesh.
-       vector_tags: List
+       element_list: List of Integers
+           A list of all specific elements in the mesh, in which the type is
+           represented by an integer.
+       scalar_tags: List of PyMOAB Tags
+           A list of all scalar tags in the mesh.
+       vector_tags: List of PyMOAB Tags
            A list of all vector tags in the mesh.
+
+    Raises:
+    _______
+       LookupError: If no element of the user specified type is found.
     """
 
     # Retrieve an arbitrary MBHEX element in the mesh and extract the tag list.
     root = mb.get_root_set()
     element_list = mb.get_entities_by_type(root, element)
 
-    # If there are none of the specified mesh elements, print a warning and exit.
+    # Warn the user if there are none of the specified mesh elements.
     if len(element_list) == 0:
-        print("WARNING: No hex elements were found in this data file.")
-        exit()
+        raise LookupError()
 
     tag_list = mb.tag_get_tags_on_entity(element_list[0])
 
-    # Check if each tag is a vector. If so, append it to the list.
+    # Check the type of each tag and append to the appropriate list.
+    scalar_tags = []
     vector_tags = []
     for tag in tag_list:
         tag_length = mb.tag_get_length(tag)
         if tag_length > 1:
             vector_tags.append(tag)
+        else:
+            scalar_tags.append(tag)
+
+    return element_list, scalar_tags, vector_tags
+
+
+def create_database(mesh_file, mb_ref, mb_exp, hexes, scal_tags, vec_tag, dir_name):
+    """
+    Expand the vector tag on each element in the given mesh data file. Write a
+    file to disk for each index with the corresponding scalar tag value.
+
+    Input:
+    ______
+       mesh_file: str
+           User supplied mesh file location.
+       mb_ref: Core
+           A PyMOAB core instance with a loaded data file for reference.
+       mb_exp: Core
+           A PyMOAB core instance with a loaded data file for expanding
+           vector tags.
+       hexes: List of Integers
+           A list of all hex elements in the mesh.
+       scal_tags: List of PyMOAB Tags
+           A list of all scalar tags in the mesh.
+       vec_tag: PyMOAB Tag
+           The vector tag to be expanded.
+       dir_name: str
+           The name of the directory in which to create and populate each
+           vector tag directory.
+
+    Returns:
+    ________
+       none
+    """
+
+    # Get the length and tag name of the vector tag.
+    length = vec_tag.get_length()
+    name = vec_tag.get_name()
+
+    # Create a directory to store the vector tag expansion files.
+    vec_dir_name = name + "_database"
+    os.mkdir(dir_name + "/" + vec_dir_name)
```

Comment:
Are there checks in place to make sure we aren't overwriting pre-existing data? We wouldn't want to overwrite previously generated data or corrupt an unrelated set of data on someone's filesystem.

I think a check at the highest level when we create the top database directory would suffice. Specifically we probably want to make sure that the directory doesn't exist or is empty. You could also add an optional `-f` flag which will ignore this check and write the database regardless to avoid annoyance at having to remove this directory every time the script is run.

Reply:
There is a check to make sure the top most directory is not overwritten in the `expand_vector_tags` function! If one with the same name exists, the appropriate integer is appended to the end. I'm not sure if a check is needed for these subdirectories, because if the top one is unique then they shouldn't be in danger of overwriting anything!

Reply:
correct - checks for the vector name subdirectories should not be necessary since the top directory is checked.